README.md
# yinka-utils
---
## Functions 
   This source can build some bins used on yinka board, 
   such as a daemon program named yinkad, we will add our 
   updater program in it soon. you can put the yinkad.conf
   to /etc/yinkad.conf like this:

    # set general delay value
    [General Sets]
    delay = 10

    # set process name and monitor status by on/off
    [yinka-terminal]
    cmdline = cmdline &
    program_name = process_name
    switch = on

## Build deps
    autotools-dev
    debhelper
    pkg-config
    shc
    libxml2-dev
    libjson-c-dev
    libcurl4-openssl-dev

## How to build
    ./autogen.sh
    ./configure --prefix=/usr
    make && sudo make install

## Build a debian package
    dpkg-buildpackage -b


ChangeLog
commit 8f1f4e81708c99a50201fd935e36f46ead94561d
Merge: b05a7a3 10ac499
Author: Steven Lee <geekerlw@gmail.com>
Date:   Mon Mar 6 13:34:34 2017 +0800

    Merge pull request #3 from liuqsqq/master
    
    add safe reboot ,force reboot and fix some formats

commit 10ac49964a3e69caeb3395008cb2a3ddd580e34b
Author: liuqs <liuqs@cqporun.com>
Date:   Mon Mar 6 11:15:16 2017 +0800

    fix some formats

commit b05a7a370c4c622ddc265707b4d6975e909f0f2e
Author: geekerlw <geekerlw@gmail.com>
Date:   Mon Mar 6 10:44:28 2017 +0800

    script: fix file cp error under running state

commit 593df59c763dbeb049166ed18ef41b57abcc41fd
Author: liuqs <liuqs@cqporun.com>
Date:   Fri Mar 3 21:18:20 2017 +0800

    add safe reboot, force reboot fuctions; modify some formats

commit a5d4074bfee43f470fcbcedd9b7681ee426bf5cb
Merge: a239384 ca95f5f
Author: liuqsqq <liuqs@cqporun.com>
Date:   Thu Mar 2 15:16:33 2017 +0800

    Merge pull request #1 from geekerlw/master
    
    merge liwen's modify

commit a239384d9c3f18e038f8628d75c37f32a3dd2fec
Author: liuqs <liuqs@cqporun.com>
Date:   Thu Mar 2 15:14:19 2017 +0800

    fix some conflicts

commit ca95f5f4ea256f3d887f6a73d259ad652d5fc306
Author: geekerlw <geekerlw@gmail.com>
Date:   Wed Mar 1 16:47:32 2017 +0800

    debian: fix package depends

commit a1f39a2fac3fc031f1c5e70000b83aadf5af4857
Author: geekerlw <geekerlw@gmail.com>
Date:   Wed Mar 1 16:29:11 2017 +0800

    debian: update to version 1.0-2ubuntu1

commit cd577c9a488e57a091ff977bd1fc212e5895ede3
Author: geekerlw <geekerlw@gmail.com>
Date:   Wed Mar 1 16:00:38 2017 +0800

    scripts: add a script to control keyboard and mouse input

commit d6dd3aa5c8887b6cd2b9d68ae122b41b7fa8ebad
Author: geekerlw <geekerlw@gmail.com>
Date:   Wed Mar 1 15:56:41 2017 +0800

    scripts: fix yinka-network wicd daemon error

commit 487de5e37f82904f5a6009a8e441c0455d50e0f0
Author: geekerlw <geekerlw@gmail.com>
Date:   Wed Mar 1 14:46:26 2017 +0800

    yinkad: rename player parole to PRAdsPlayer

commit cbe023bf053b3cbbb3cda280c57f706136d00875
Author: geekerlw <geekerlw@gmail.com>
Date:   Wed Mar 1 14:32:38 2017 +0800

    updater: add a systemd service file for updater program

commit 47a5398d1aa72e6b0102b1d782832ed66c83248f
Author: geekerlw <geekerlw@gmail.com>
Date:   Wed Mar 1 14:16:35 2017 +0800

    build: mv yinkad.service to data dir

commit da1968b4da1470641b5f1e24d04c5ed652482b3a
Author: geekerlw <geekerlw@gmail.com>
Date:   Wed Mar 1 13:21:22 2017 +0800

    update README.md

commit b4b1adc260089dfac9b6a17b6bf6d151bbe5862e
Author: geekerlw <geekerlw@gmail.com>
Date:   Wed Mar 1 12:53:53 2017 +0800

    build: use macros instead of manual sets

commit 11a83ea5453ee595460c0feb41e7ea8465e6834a
Author: geekerlw <geekerlw@gmail.com>
Date:   Wed Mar 1 12:53:12 2017 +0800

    updater: add an auto update tool

commit 31e578074458fb6b62c3c79f28b75e593e8daedd
Author: geekerlw <geekerlw@gmail.com>
Date:   Wed Mar 1 11:52:14 2017 +0800

    debian: release version 1.0-1ubuntu2

commit 090080688e27118cd0ef34cf93e884c5895e6f38
Author: geekerlw <geekerlw@gmail.com>
Date:   Tue Feb 28 17:14:45 2017 +0800

    scripts: fix yinka-network.sh error '\r'

commit 937b608ad0b42b684e5c105f9aec75f00c14376a
Author: geekerlw <geekerlw@gmail.com>
Date:   Tue Feb 28 17:13:14 2017 +0800

    build: change yinkad.conf to data dir

commit c6c9b3c7f9aaa7338de9c1a57fbad853305c5ea7
Author: geekerlw <geekerlw@gmail.com>
Date:   Tue Feb 28 16:08:31 2017 +0800

    build: update configure.ac to check shell tool

commit e3db818499d3ebba7c67c04026f9e962c2f67b1e
Author: geekerlw <geekerlw@gmail.com>
Date:   Tue Feb 28 15:16:34 2017 +0800

    debian: prepare version 1.0-1ubuntu2 build

commit b71f4013d90729feb4d7c8b283c4418d0a5c9982
Author: geekerlw <geekerlw@gmail.com>
Date:   Tue Feb 28 15:11:22 2017 +0800

    build: add build rules for scripts

commit c594995116ba2ae4b51508f576eb71e30fae41fd
Author: geekerlw <geekerlw@gmail.com>
Date:   Tue Feb 28 15:10:09 2017 +0800

    scripts: add scripts update printer_set etc.

commit f4357073a875c3373d1b06c0447163b5884c284b
Author: geekerlw <geekerlw@gmail.com>
Date:   Tue Feb 28 14:47:32 2017 +0800

    debian: release version 1.0-1ubuntu1

commit 01a27b0844f8b3fe46e9422cb41905b189c862f4
Author: geekerlw <geekerlw@gmail.com>
Date:   Wed Feb 22 16:22:43 2017 +0800

    debian: update to 1.0-1ubuntu1

commit dd3630b22a798e506e24e5145397ee463a5de754
Author: geekerlw <geekerlw@gmail.com>
Date:   Wed Feb 22 16:04:32 2017 +0800

    write commit log into ChangeLog

commit bee98d352e79c62b40d1e02db8b6222709d45d59
Author: geekerlw <geekerlw@gmail.com>
Date:   Wed Feb 22 14:28:38 2017 +0800

    fix the file open error
    once open the file error, the *fp won't close

commit 1f31e54fd61dc197b751e300ff6f44bc7c073390
Author: geekerlw <geekerlw@gmail.com>
Date:   Mon Feb 20 13:50:50 2017 +0800

    change config file path to /etc/yinka

commit f895569fa17ecfbb6f723da9bd1577a97480f07f
Author: geekerlw <geekerlw@gmail.com>
Date:   Mon Feb 20 13:43:57 2017 +0800

    rename yinka-terminal and ads

commit e9f9bbc8d255ef447a8b09e1887319150dbff25e
Author: geekerlw <geekerlw@gmail.com>
Date:   Mon Feb 20 13:01:13 2017 +0800

    clean up the code merge from liuqs

commit 88e49a5cbafe676319a8b341977b9a8dccf424ef
Author: liuqsqq <liuqs@cqporun.com>
Date:   Mon Feb 20 11:39:28 2017 +0800

    modify some formats (#2)
    
    * 	modified:   yinkad.c
    	modified:   yinkad.h
    
    * test
    
    * test del
    
    * modify some formats

commit 023440fd2101eb570292fcbbb9a19aeb18cd3196
Author: liuqs <liuqs@cqporun.com>
Date:   Mon Feb 20 09:26:34 2017 +0800

    modify some formats

commit 2fe5587c08fe105f552b8fab3bbb78d51498877e
Author: geekerlw <geekerlw@gmail.com>
Date:   Mon Feb 20 09:14:09 2017 +0800

    modified yinkad.service to match debian package install path

commit 1120757980d073398e07539e8a1865c94d1a4f25
Author: liuqs <liuqs@cqporun.com>
Date:   Sat Feb 18 17:15:08 2017 +0800

    test del


src/utils/md5.c
/*
 *   Copyright (C) 2017  Steven Lee <geekerlw@gmail.com>
 *
 *   This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "md5.h"  
#include <memory.h>  
   
unsigned char PADDING[] = {  
    0x80,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,  
    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,  
    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,  
    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0  
};  

void MD5Init(MD5_CTX *context) 
{  
    context->count[0] = 0;  
    context->count[1] = 0;  
    context->state[0] = 0x67452301;  
    context->state[1] = 0xEFCDAB89;  
    context->state[2] = 0x98BADCFE;  
    context->state[3] = 0x10325476;  
}  

void MD5Update(MD5_CTX *context, unsigned char *input, unsigned int inputlen)  
{  
    unsigned int i = 0;  
    unsigned int index = 0;  
    unsigned int partlen = 0;  

    index = (context->count[0] >> 3) & 0x3F;  
    partlen = 64 - index;  
    context->count[0] += inputlen << 3;  

    if(context->count[0] < (inputlen << 3))  
     context->count[1]++;  
    context->count[1] += inputlen >> 29;  

    if(inputlen >= partlen){  
        memcpy(&context->buffer[index], input,partlen);  
        MD5Transform(context->state, context->buffer);  

        for(i = partlen; i+64 <= inputlen; i+=64)  
            MD5Transform(context->state, &input[i]);  

        index = 0;          
    }    
    else {  
        i = 0;  
    }  
    memcpy(&context->buffer[index], &input[i], inputlen-i);  
}  

void MD5Final(MD5_CTX *context, unsigned char digest[16])  
{  
    unsigned int index = 0,padlen = 0;  
    unsigned char bits[8];  

    index = (context->count[0] >> 3) & 0x3F;  
    padlen = (index < 56)?(56-index):(120-index);  
    MD5Encode(bits, context->count, 8);  
    MD5Update(context, PADDING, padlen);  
    MD5Update(context, bits, 8);  
    MD5Encode(digest, context->state, 16);  
}  

void MD5Encode(unsigned char *output,unsigned int *input,unsigned int len)  
{  
    unsigned int i = 0;  
    unsigned int j = 0;  

    while(j < len) {  
     output[j] = input[i] & 0xFF;    
     output[j+1] = (input[i] >> 8) & 0xFF;  
     output[j+2] = (input[i] >> 16) & 0xFF;  
     output[j+3] = (input[i] >> 24) & 0xFF;  
     i++;  
     j += 4;  
    }  
}  

void MD5Decode(unsigned int *output, unsigned char *input, unsigned int len)  
{  
    unsigned int i = 0;  
    unsigned int j = 0;  

    while(j < len) {  
     output[i] = (input[j]) |  
         (input[j+1] << 8) |  
         (input[j+2] << 16) |  
         (input[j+3] << 24);  
     i++;  
     j += 4;   
    }  
}  

void MD5Transform(unsigned int state[4], unsigned char block[64])  
{  
    unsigned int a = state[0];  
    unsigned int b = state[1];  
    unsigned int c = state[2];  
    unsigned int d = state[3];  
    unsigned int x[64];  

    MD5Decode(x,block,64);  

    FF(a, b, c, d, x[ 0], 7, 0xd76aa478); /* 1 */  
    FF(d, a, b, c, x[ 1], 12, 0xe8c7b756); /* 2 */  
    FF(c, d, a, b, x[ 2], 17, 0x242070db); /* 3 */  
    FF(b, c, d, a, x[ 3], 22, 0xc1bdceee); /* 4 */  
    FF(a, b, c, d, x[ 4], 7, 0xf57c0faf); /* 5 */  
    FF(d, a, b, c, x[ 5], 12, 0x4787c62a); /* 6 */  
    FF(c, d, a, b, x[ 6], 17, 0xa8304613); /* 7 */  
    FF(b, c, d, a, x[ 7], 22, 0xfd469501); /* 8 */  
    FF(a, b, c, d, x[ 8], 7, 0x698098d8); /* 9 */  
    FF(d, a, b, c, x[ 9], 12, 0x8b44f7af); /* 10 */  
    FF(c, d, a, b, x[10], 17, 0xffff5bb1); /* 11 */  
    FF(b, c, d, a, x[11], 22, 0x895cd7be); /* 12 */  
    FF(a, b, c, d, x[12], 7, 0x6b901122); /* 13 */  
    FF(d, a, b, c, x[13], 12, 0xfd987193); /* 14 */  
    FF(c, d, a, b, x[14], 17, 0xa679438e); /* 15 */  
    FF(b, c, d, a, x[15], 22, 0x49b40821); /* 16 */  

    /* Round 2 */  
    GG(a, b, c, d, x[ 1], 5, 0xf61e2562); /* 17 */  
    GG(d, a, b, c, x[ 6], 9, 0xc040b340); /* 18 */  
    GG(c, d, a, b, x[11], 14, 0x265e5a51); /* 19 */  
    GG(b, c, d, a, x[ 0], 20, 0xe9b6c7aa); /* 20 */  
    GG(a, b, c, d, x[ 5], 5, 0xd62f105d); /* 21 */  
    GG(d, a, b, c, x[10], 9,  0x2441453); /* 22 */  
    GG(c, d, a, b, x[15], 14, 0xd8a1e681); /* 23 */  
    GG(b, c, d, a, x[ 4], 20, 0xe7d3fbc8); /* 24 */  
    GG(a, b, c, d, x[ 9], 5, 0x21e1cde6); /* 25 */  
    GG(d, a, b, c, x[14], 9, 0xc33707d6); /* 26 */  
    GG(c, d, a, b, x[ 3], 14, 0xf4d50d87); /* 27 */  
    GG(b, c, d, a, x[ 8], 20, 0x455a14ed); /* 28 */  
    GG(a, b, c, d, x[13], 5, 0xa9e3e905); /* 29 */  
    GG(d, a, b, c, x[ 2], 9, 0xfcefa3f8); /* 30 */  
    GG(c, d, a, b, x[ 7], 14, 0x676f02d9); /* 31 */  
    GG(b, c, d, a, x[12], 20, 0x8d2a4c8a); /* 32 */  

    /* Round 3 */  
    HH(a, b, c, d, x[ 5], 4, 0xfffa3942); /* 33 */  
    HH(d, a, b, c, x[ 8], 11, 0x8771f681); /* 34 */  
    HH(c, d, a, b, x[11], 16, 0x6d9d6122); /* 35 */  
    HH(b, c, d, a, x[14], 23, 0xfde5380c); /* 36 */  
    HH(a, b, c, d, x[ 1], 4, 0xa4beea44); /* 37 */  
    HH(d, a, b, c, x[ 4], 11, 0x4bdecfa9); /* 38 */  
    HH(c, d, a, b, x[ 7], 16, 0xf6bb4b60); /* 39 */  
    HH(b, c, d, a, x[10], 23, 0xbebfbc70); /* 40 */  
    HH(a, b, c, d, x[13], 4, 0x289b7ec6); /* 41 */  
    HH(d, a, b, c, x[ 0], 11, 0xeaa127fa); /* 42 */  
    HH(c, d, a, b, x[ 3], 16, 0xd4ef3085); /* 43 */  
    HH(b, c, d, a, x[ 6], 23,  0x4881d05); /* 44 */  
    HH(a, b, c, d, x[ 9], 4, 0xd9d4d039); /* 45 */  
    HH(d, a, b, c, x[12], 11, 0xe6db99e5); /* 46 */  
    HH(c, d, a, b, x[15], 16, 0x1fa27cf8); /* 47 */  
    HH(b, c, d, a, x[ 2], 23, 0xc4ac5665); /* 48 */  

    /* Round 4 */  
    II(a, b, c, d, x[ 0], 6, 0xf4292244); /* 49 */  
    II(d, a, b, c, x[ 7], 10, 0x432aff97); /* 50 */  
    II(c, d, a, b, x[14], 15, 0xab9423a7); /* 51 */  
    II(b, c, d, a, x[ 5], 21, 0xfc93a039); /* 52 */  
    II(a, b, c, d, x[12], 6, 0x655b59c3); /* 53 */  
    II(d, a, b, c, x[ 3], 10, 0x8f0ccc92); /* 54 */  
    II(c, d, a, b, x[10], 15, 0xffeff47d); /* 55 */  
    II(b, c, d, a, x[ 1], 21, 0x85845dd1); /* 56 */  
    II(a, b, c, d, x[ 8], 6, 0x6fa87e4f); /* 57 */  
    II(d, a, b, c, x[15], 10, 0xfe2ce6e0); /* 58 */  
    II(c, d, a, b, x[ 6], 15, 0xa3014314); /* 59 */  
    II(b, c, d, a, x[13], 21, 0x4e0811a1); /* 60 */  
    II(a, b, c, d, x[ 4], 6, 0xf7537e82); /* 61 */  
    II(d, a, b, c, x[11], 10, 0xbd3af235); /* 62 */  
    II(c, d, a, b, x[ 2], 15, 0x2ad7d2bb); /* 63 */  
    II(b, c, d, a, x[ 9], 21, 0xeb86d391); /* 64 */  
    state[0] += a;  
    state[1] += b;  
    state[2] += c;  
    state[3] += d;  
}
